executor/src/witgen/machines/block_machine.rs

```diff
@@ -315,167 +319,83 @@ impl<'a, T: FieldElement> BlockMachine<'a, T> {
+            block,
+            identity_processor,
+            &self.identities,
+            self.fixed_data,
+            self.row_factory.clone(),
+            &self.witness_cols,
+        )
+        .with_calldata(Calldata::new(left.to_vec(), right));
 
-        log::trace!("End processing block machine");
+        let outer_assignments = processor.solve(&mut sequence_iterator)?;
+        let (new_block, left_new) = processor.finish();
+        let left_new = left_new.unwrap();
 
         // Only succeed if we can assign everything.
         // Otherwise it is messy because we have to find the correct block again.
-        let success = left_mut.iter().all(|v| v.is_constant());
+        let success = left_new.iter().all(|v| v.is_constant());
 
         if success {
+            log::trace!("End processing block machine (successfully)");
+            self.append_block(new_block)?;
+
             // We solved the query, so report it to the cache.
             self.processing_sequence_cache
-                .report_processing_sequence(left, processing_sequence_iterator);
-            Ok(outer_assignments)
-        } else if !errors.is_empty() {
-            Err(errors
-                .into_iter()
-                .reduce(|x: EvalError<T>, y| x.combine(y))
-                .unwrap())
+                .report_processing_sequence(left, sequence_iterator);
+            Ok(EvalValue::complete(outer_assignments))
         } else {
+            log::trace!("End processing block machine (incomplete)");
             Ok(EvalValue::incomplete(
                 IncompleteCause::BlockMachineLookupIncomplete,
             ))
         }
     }
 
-    fn compute_updates(
-        &self,
-        row: DegreeType,
-        fixed_data: &'a FixedData<T>,
-        left: &[AffineExpression<&'a PolynomialReference, T>],
-        right: &'a SelectedExpressions<T>,
-        identity: IdentityInSequence,
-        identity_processor: &mut IdentityProcessor<'a, '_, T>,
-    ) -> EvalResult<'a, T> {
-        match identity {
-            IdentityInSequence::Internal(i) => identity_processor.process_identity(
-                self.identities[i],
-                &self.get_current_row_pair(row, fixed_data),
-            ),
-            IdentityInSequence::OuterQuery => {
-                assert!(row as usize % self.block_size == self.block_size - 1);
-                identity_processor.process_link(
-                    left,
-                    right,
-                    &self.get_current_row_pair(row, fixed_data),
-                )
-            }
-        }
-    }
-
-    fn apply_updates(
-        &mut self,
-        row: DegreeType,
-        identity: IdentityInSequence,
-        updates: &EvalValue<&'a PolynomialReference, T>,
-        left_mut: &mut [AffineExpression<&'a PolynomialReference, T>],
-    ) -> bool {
-        if updates.constraints.is_empty() {
-            return false;
+    /// Takes a block of rows, which contains the last row of the previous block
```

Comment:
Can you clarify if `new_block` is the "next block" or if the next block is the one after `new_block`? (In the comment, not in the review)